[Guided onboarding] Security tour

yuliacech opened the pull request:
## Summary
Addresses https://github.com/elastic/cloud/issues/102847

This PR adds an onboarding tour for the Security solution. A clickable design prototype is available under this [link](https://ela.st/onboarding-prototype) (select v 8.4). Related to a similar onboarding tour for Observability (https://github.com/elastic/kibana/pull/133909).
Tests will be added as a follow up PR (tracked in https://github.com/elastic/cloud/issues/104482).

### Screenshots 
<img width="1469" alt="Screenshot 2022-07-20 at 14 28 36" src="https://user-images.githubusercontent.com/6585477/179993001-b0af9011-89fb-4867-88bf-d71729f1ca73.png">


<details>

<img width="1469" alt="Screenshot 2022-07-20 at 14 28 42" src="https://user-images.githubusercontent.com/6585477/179993297-4281ed7f-2557-41a1-8a7d-a4dc2e0ed247.png">

<img width="1469" alt="Screenshot 2022-07-20 at 14 28 50" src="https://user-images.githubusercontent.com/6585477/179993541-910e87fe-5fa4-48a6-8341-d011b489b47c.png">

<img width="1469" alt="Screenshot 2022-07-20 at 14 29 03" src="https://user-images.githubusercontent.com/6585477/179993670-f91251f5-4261-4ec9-812e-2731eb90b3cc.png">

<img width="1469" alt="Screenshot 2022-07-20 at 14 29 17" src="https://user-images.githubusercontent.com/6585477/179993847-c8551a74-61b4-4a5e-9e80-1f0fad44c97e.png">


</details>

### Screencast

https://user-images.githubusercontent.com/6585477/179992524-b90a1998-0111-48be-bcd7-cac192f38cdc.mov





### How to test
1. Navigate to Stack Management -> Advanced Settings and enable the new Security navigation (set `securitySolution:enableGroupedNav` to ON)
2. Navigate to /home#/getting_started and select Security
3. Check that the tour is correctly displayed: 
  - The tour can be skipped.
  - The tour can be completed.
  - When the tour is displayed, the side navigation can't be collapsed. 
  - The tour is not displayed on small screens (xs and s breakpoints).
  - The tour is displayed even if `hideAnnoucements` setting is ON, because those onboarding tours are activated by the user action and not displayed by default. 


### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))

yuliacech committed:
[Guided onboarding] Initial security tour setup

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

yuliacech committed:
[Guided onboarding][Security tour] Use `canBeCollapsed` prop of the side nav

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
[Guided onboarding] Changed the localStorage key for the tour activation

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
[Guided onboarding] Added "hideAnnouncements" check, added screen size check, updated copy

yuliacech committed:
[Guided onboarding] Added data-test-subj to buttons for fullstory tracking

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
[Guided onboarding] Added i18n to the tour config

yuliacech committed:
[Guided onboarding] Added gifs to the tour

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
[Guided onboarding] Commented out gif import errors for now

yuliacech committed:
[Guided onboarding] Updated a jest test snapshot

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
[Guided onboarding] Removed testing buttons from the landing cards

yuliacech committed:
[Guided onboarding] Deleted unnecessary properties in the tour context

yuliacech committed:
[Guided onboarding] Updated to the correct `repositionOnScroll` value

yuliacech commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
[Guided onboarding] Deleted ts-error after gif modules have been added

yuliacech committed:
[Guided onboarding] Updated import

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into guided_onboarding/security

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

yuliacech committed:
[Guided onboarding] Removed `hideAnnouncements` check

yuliacech committed:
Update x-pack/plugins/security_solution/public/common/components/guided_onboarding/tour_config.ts

Co-authored-by: gchaps <33642766+gchaps@users.noreply.github.com>

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

YulNaumenko commented on x-pack/plugins/security_solution/public/common/components/navigation/use_security_solution_navigation/use_primary_navigation.tsx:
```diff
@@ -51,6 +54,7 @@ export const usePrimaryNavigation = ({
   });
 
   return {
+    canBeCollapsed: !isTourShown,
```
If it is possible could you please add some unit test coverage to check that behavior?

alisonelizabeth commented on x-pack/plugins/security_solution/public/common/components/guided_onboarding/tour.tsx:
```diff
@@ -0,0 +1,182 @@
+const SECURITY_TOUR_ACTIVE_KEY = 'guidedOnboarding.security.tourActive';
+const SECURITY_TOUR_STEP_KEY = 'guidedOnboarding.security.tourStep';
+const getIsTourActiveFromLocalStorage = (): boolean => {
+  return Boolean(JSON.parse(String(localStorage.getItem(SECURITY_TOUR_ACTIVE_KEY))));
```
Is it necessary to convert this to a string, and then parse it? 🤔

yuliacech replied:
No, you're right @alisonelizabeth , it's some older code. I removed it in https://github.com/elastic/kibana/pull/132641/commits/21b7cce5efa679080338cf3679320ead383e5911

alisonelizabeth commented on x-pack/plugins/security_solution/public/common/components/guided_onboarding/tour.tsx:
```diff
@@ -0,0 +1,182 @@
+  return tourConfig.map((stepConfig) => {
+    const { content, imageConfig, ...rest } = stepConfig;
+    return (
+      // @ts-expect-error
```
Can you explain why this is needed?

yuliacech replied:
Good catch, @alisonelizabeth, I removed the error and fixed the config in this [commit](https://github.com/elastic/kibana/pull/132641/commits/763249e87a8a9bcb7c26a8070b7a5d7437e71814).

alisonelizabeth commented on x-pack/plugins/security_solution/public/common/components/guided_onboarding/tour.tsx:
```diff
@@ -0,0 +1,182 @@
+};
+
+const getTourStepFromLocalStorage = (): number => {
+  return Number(JSON.parse(String(localStorage.getItem(SECURITY_TOUR_STEP_KEY) ?? 1)));
```
Same here

alisonelizabeth reviewed (commented):
Nice job! Tested and everything worked as expected. I left a couple minor comments in the code.

I think we should also end the tour if the user clicks the "Add integrations" button. Similar to https://github.com/elastic/kibana/issues/135927. It's not as bad, since it's a separate page that the user is navigated to. However, if the user came back to security, it would be a little weird for the last step in the tour to still display.

yuliacech committed:
[Guided onboarding] Added tests for security nav hook and tour

yuliacech committed:
[Guided onboarding] Removed TS error ignore

yuliacech committed:
[Guided onboarding] Removed unneeded string casting for local storage items

yuliacech commented:
Thanks a lot for the review, @YulNaumenko! I added some unit tests, would you mind having another look please?

yuliacech commented:
Thanks a lot for the review, @alisonelizabeth! I addressed your comments in the code, would you mind having another look please? 
That's a good point about the last step and ending the tour by clicking the "Add data" button. I'll make sure to include it into the follow up PR with Cypress tests, that I'm working on right now.

alisonelizabeth reviewed (approved):
Thanks for addressing my feedback! Latest LGTM.

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech commented:
Hi @gchaps, thanks a lot for your review! Just wanted to make sure, is this PR good to go from your perspective?

gchaps reviewed (approved):
Text LGTM

adamwdraper commented:
@yuliacech Before we merge this, we need to align with the Security Design Team.  For context, we are turning on the new navigation by default in 8.4, and this tour is not aligned with that direction.  I will reach out to you via slack to follow up on a time to discuss.

adamwdraper reviewed (changes requested):
We will need to sync with Security Design before moving forward on this pr.  If we are building a tour it needs to support the new navigation that is the default for 8.4.

yuliacech committed:
[Guided onboarding] Updating id for Security header in the nav

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

machadoum commented:
Hi @yuliacech, Sorry for jumping in, but did you test this feature with the new security solution menu?
You can enable it on `Stack Management / Advanced Settings / New streamlined navigation`.

yuliacech committed:
[Guided onboarding] Removing unnecessary imports from a merge conflict

yuliacech commented:
Thanks a lot for calling that out, @adamwdraper! I was not aware of the new side nav and that it will be on by default in 8.4. I think most of the steps could work with the new design, except for "Rules". Has that nav item been moved to a sub menu somewhere?

yuliacech commented:
Thanks a lot for the tip, @machadoum! I didn't know we can enable the new nav already. I'll try it out

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
[Guided onboarding] Added a tour config for the Security grouped navigation

gavinwye commented:
Late to the party here. I'm leading design for the guided onboarding for security and this PR seems to contradict that work. 

This tour guides people around the product and then points them to integrations. That's the opposite of the process we're designing for Security onboarding.

With the new navigation, I'm not sure we need a tour like this, the nav is self-explanatory. 

In 8.6 when we release the guided onboarding we'll probably (still doing the design work) have tours that:
1. introduce the alert triage process 
2. introduce cases and investigations 
3. help users to do more with security like looking at dashboards and reports, and inviting team members.

If we put this release this tour in 8.4 I expect we'll be removing it again in 8.6. 

I don't know why this needs to be released in 8.4 but I don't think we should do that knowing that we'll change it in 8.6.

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

adamwdraper commented:
For reference, here is an document with updated text and the associated links for the tour for the new navigation: https://docs.google.com/document/d/13iJRPh9B44U7umytZCoRd3Z4t-f6qethq4aba9q_WHs/edit?usp=sharing

Most text was taken directly from existing documentation to ensure messaging alignment.

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
[Guided onboarding] Updated tour copy

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

yuliacech committed:
[Guided onboarding] Deleted tour for the old nav

yuliacech commented:
Hi @gavinwye, I added support for the new Security side nav, could you please have a look?

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

yuliacech committed:
Merge branch 'main' into guided_onboarding/security

YulNaumenko reviewed (approved):
LGTM! Thank you for adding the tests!

yuliacech committed:
Merge branch 'main' into guided_onboarding/security
